tests/src/test/scala/com/nvidia/spark/rapids/QueryFuzzer.scala

```diff
@@ -0,0 +1,240 @@
+//      if (rand.nextFloat() > 0.7) {
+//        return generateDataSource(spark)
+//      } else {
+//        return renameColumns(
+//          spark.read.parquet("/mnt/tpcds/parquet/sf100-decimals/web_sales.dat"))
+//      }
+    }
+
+    // recurse down first to get a DataFrame
+    val df = randomOperator(spark, depth+1, maxDepth)
+
+    // build a list of available transformations
+    val ctx = FuzzContext(spark, this)
+    val transformations = Seq(
+      Filter(ctx),
+      SortRandomColumns(ctx),
+      RandomCastFromString(ctx),
+      Aggregate(ctx),
+      Repartition(ctx)
+    )
+
+    // filter down to a list of transformations that are applicable in the current context
+    val validTransformations = transformations.filter(_.canTransform(df))
+    if (validTransformations.isEmpty) {
+      df
+    } else {
+      // pick a transformation at random
+      val tf = randomElement(validTransformations)
+      // apply the transformation
+      tf.transform(df)
+    }
+  }
+
+  /**
+   * Generate a random in-memory DataFrame.
+   */
+  def generateDataSource(spark: SparkSession) = {
+    val numFields = 1 + rand.nextInt(32)
+    val supportedTypes = Seq(DataTypes.IntegerType,
+      DataTypes.StringType,
+      new DecimalType(10,2),
+      DataTypes.DateType,
+      DataTypes.TimestampType,
+      DataTypes.BooleanType,
+      DataTypes.DoubleType,
+      DataTypes.FloatType)
+    val dataTypes = (0 until numFields).map(_ => randomElement(supportedTypes))
+    val schema = FuzzerUtils.createSchema(dataTypes)
+    val options = FuzzerOptions(validStringChars = Some(" \t\r\n0123456789.+-/:aidfnT"),
+      maxStringLen = 12)
+    val df = FuzzerUtils.generateDataFrame(spark, schema, rowCount = 256, options, seed)
+      .repartition(1 + rand.nextInt(4))
+    renameColumns(df)
+  }
+
+  /**
+   * Pick a random element from a sequence.
+   */
+  def randomElement[T](seq: Seq[T]): T = {
+    seq(rand.nextInt(seq.length))
+  }
+
+  /**
+   * Rename all of the columns in a DataFrame so that they are unique.
+   */
+  def renameColumns(df: DataFrame): DataFrame = {
+    var dfRenamed = df;
+    for (name <- df.columns) {
+      dfRenamed = dfRenamed.withColumnRenamed(name, nextName())
+    }
+    dfRenamed
+  }
+
+  /**
+   * Generate the next column name.
+   */
+  def nextName(): String = "c" + idGen.getAndIncrement()
+
+}
+
+case class FuzzContext(spark: SparkSession, fuzzer: QueryFuzzer)
+
+trait Transformation {
+  def canTransform(df: DataFrame): Boolean
+  def transform(df: DataFrame): DataFrame
+}
+
+case class Filter(ctx: FuzzContext) extends Transformation {
+
+  override def canTransform(df: DataFrame): Boolean = {
+    df.schema.fields.exists(_.dataType == DataTypes.IntegerType)
+  }
+
+  override def transform(df: DataFrame): DataFrame = {
+    val intCols = df.schema.fields.filter(_.dataType == DataTypes.IntegerType)
+    df.filter(col(intCols(0).name).gt(col(intCols(1).name)))
```

Comment:
I assume the plan is to make this a bit more sophisticated in the future, with RNG-selected operations, columns, and column ordering so it's not always lower-indexed columns being compared against higher-indexed columns.

Reply:
Yes, all of this is just a slightly cleaned-up version of the demo that I hacked together and is just a starting point.